### ISSUE TITLE
Is it really local??

### ISSUE THREAD

**Issue by rajat2485** [rospogrigio/localtuya#539]
Purpose of this integration will be once devices are ported from Smartife or tuya smart app, we no longer need these app and can remove devices from them - I tried. the integration broke. Is my understanding correct - as we keep apps, we are still sharing our device and information over with these apps

**Comment by patrik-malina** [comment]
> Purpose of this integration will be once devices are ported from Smartife or tuya smart app, we no longer need these app and can remove devices from them - I tried. the integration broke. Is my understanding correct - as we keep apps, we are still sharing our device and information over with these apps

As I understand, when you remove device from app, it lose/reset its localkey, so that is why it no working anymore.

Although if you leve device registered in app, even if you delete app, it is working.

**Comment by rajat2485** [comment]
Ok, Thanks Patrik. So once registered and setup  are we saying that communication is local and not routed via IOT cloud or smartlife

**Comment by vmartinv** [comment]
@rajat2485  If you want to delete the devices I guess you first need to cut all communication to the internet of the device. I did this by accessing the options in my router (it can block internet for certain IPs) and also blocking all tuya domains in my DNS settings. Now when I go to the IoT developer dashboard it shows all my devices are offline, but you need to block the access.

**Comment by tsightler** [comment]
Basically all of these tools which talk to Tuya devices locally take advantage of a feature that Tuya built into their devices that allows their app to control them even when the internet is offline/unreachable, as long as the device running the Tuya/Smartlife app and the devices themselves are on the same network.  This is actually a pretty cool feature, once you setup the Tuya bulbs in the smartlife app you can literally unplug the internet connection on your home wifi and, after a few seconds/minutes, the Tuya app will still be able to control the lights locally.

In the early days with Tuya firmware that used the 3.1 protocol version this local traffic was not even encrypted, but almost all devices these days use Tuya 3.3 protocol which generates a unique localkey during initial registration to Tuya cloud which the app then uses to encrypt all local communications between the app and the device.  If you unregister the devices from the Tuya cloud, these keys are reset, so you lose the ability to control them locally. 

As mentioned above, the "fix", if you truly want only local control, is to register them with Tuya, get the keys, and then block the devices from accessing the internet.  This way the device thinks it is still registered and will keep the local keys and local control will still work.

**Comment by tsightler** [comment]
It's possible @germanotero is referring to things like energy consumption data, which requires a constant polling from the controlling app.  See https://github.com/rospogrigio/localtuya/issues/87 and https://github.com/tuya/tuya-home-assistant/issues/211 for more details on that.  This is a well known issue with Tuya devices and local control, to get updated data from those device the application must send constant dp_refresh commands.  This could of course be implemented in localtuya, but I don't think it has been to this point.

**Comment by AdmiralStipe** [comment]
@vmartinv , just an additional info after a few days use - it works like a charm, but it has a downside too... HA database is growing like mad - state_changed() events from the plugs are happening and recording to database every second or two (which seems logical, as my eight plugs are polled every 10 seconds, so in average a little less than one per second). 
So, my database was roughly 70MB (purged on a daily basis so only last 7 days of data remained in it) before this change and now it's about 200MB and still growing. 
So I agree with Leatherface75, an option for adjusting interval instead of direct heartbeat would be nice (only I would increase it to 30 or even 60 seconds, as it's OK for my needs :) ). 

![db](https://user-images.githubusercontent.com/64564398/131489919-338f19f9-ef60-4ca9-ac39-58820ffb1b2c.jpg)

**Comment by vmartinv** [comment]
@AdmiralStipe Thanks for flagging this up. I updated the code to only save the state on changes. Please leave any other feedback in the pull request so it's easier to follow up.

**Comment by dyde121** [comment]
> the "fix", if you truly want only local control, is to register them with Tuya, get the keys, and then block the devices from accessing the internet. This way the device thinks it is still registered and will keep the local keys and local control will still work.

Even then, it doesn't seem to work.
I have they key and can connect it to local Tuya, but then if I block internet access (Including DNS) the device will not respond.
(It can still be discovered by local tuya, but not added, or if already added, cannot be interacted with)

**Comment by sibowler** [comment]
Hi @dyde121 - I had the same problem. I found that when i forced the DNS (via the individual DHCP lease) to be an external IP (i.e. 8.8.8.8), then the device responds again. It seems to recognize that it should be able to access a local DNS server, and if it can't access it, then it goes into lockdown.

**Comment by dyde121** [comment]
@sibowler Did you allow it to resolve DNS via that external server?
Or DNS resolution still blocked, just the act of giving it an external DNS server that it can’t reach was enough?

**Comment by sibowler** [comment]
No, dns resolution was still blocked. It was enough to give it an external dns server.

**Comment by sibowler** [comment]
I should also note though, that for some devices this works perfectly. Others it only allows me to turn them on/off, but not access any of the power attributes.

**Comment by skallberg** [comment]
I also have issues with energy reports.
I have the device in a seperate VLAN with routing rules allowing the device to talk to my HA server.
It works when I allow internet access for the device but when I block internet for the device I'm only able to turn on/off but I get no energy data.

**Comment by sjoerdtimmer** [comment]
This (control works but power reporting is broken) also seems to be the case for Hombli (powered by tuya) smart plugs. I have blocked all outgoing traffic as well as dns requests to the router as describe above.
